Verify that already loaded JMockit version is the expected version

Vampire opened the pull request:
This PR verifies that if a JMockit agent is already loaded in the current JVM, that it is the version expected. This prevents e. g. if you start a JVM with `-javaagent:jmockit-1.8.jar` but have `jmockit-1.24.jar` in your classpath to run your tests with that incompatibilites arise.

This also covers the case that you load JMockit in an application server to run integration tests from a branch using JMockit 1.8 and then run on the same application server instance integration tests from a branch using JMockit 1.24 without restarting the application server.

rliesenfeld commented:
Needs reproducer test first.

rliesenfeld commented:
Not specifically about this PR, but I read that arquillian.xml accepts a "jvmArgs" entry where "-javaagent" can be specified. Wouldn't that provide a clean way to activate JMockit in an Arquillian-based test suite, avoiding a lot of problems that only occur when JMockit is loaded through the Attach API?

Vampire commented:
I added a test for this.

`jvmArgs` in `arquillian.xml` will probably help if you use managed container with Arquillian (Arquillian starts and stops the container). But if you use remote container (Arquillian just deploys to an already running container) I don't think `jvmArgs` in `arquillian.xml` will have any effect. Thanks for the notice though.

Besides that, though I found and fixed these issues during trying to make our tests work nicely with Arquillian, all my PRs up to now were are independent of Arquillian and are valid for stand-alone JMockit usage too. :-)

rliesenfeld commented:
Ok, good point about the remote container scenario. And this should be the case with the "managed" scenario as well, where Arquillian starts a local container as a separate process. In such cases, though, there shouldn't be any jmockit.jar in the classpath beyond the one thats gets added to the Arquillian deployment jar. So, the need for a version check seems unlikely.

I am still considering this, but one detail that would need to change in the PR is to replace that "version.properties" file with a simple constant string value in the VERSION field.

Vampire commented:
Actually I also think that would be better, I just didn't know how to do a replace in a source file in Maven but did know how to do it with resources.

Vampire commented:
As I said, this version check is not only helpful for the Arquillian use-case, but also for the standalone JMockit use-case as described in the original comment. If you somehow configured your JVM to start with `jmockit-1.8.jar` via batch file, or maven POM, or webstart configuration or whatever and then update your project dependencies to use JMockit 1.24, you instantly get notified that there is a version conflict and you can sort it out. Without the version check, you might get much more subtle errors.

rliesenfeld commented:
Oh, no need for a source file replace through Maven. I already do manual version replacement (on HTML pages too) on every release, but it's easy, IntelliJ does it for me.

Vampire commented:
Oh, ok, I read your comment too late and now already learned how to do source file replacements correctly, so now you have a version where the project version gets injected into the class to be there as constant field at runtime. If you still prefer to just have it hard-coded and do the manual replacement, I can also change it to that of course.

rliesenfeld commented:
I imagine use of "templating-maven-plugin" in the pom.xml file is not going to work when compiling the code with the Java IDE (IntelliJ, in my case). But apart from that, I think the hard-coded version number in the .java file is a simpler solution; manual replacement is going to be done anyway as this number appears in other places.

Vampire commented:
I also use IJ, as it is the best Java IDE under the sun. :-)
And it has a button in the Maven tool window:
![generate-sources](https://cloud.githubusercontent.com/assets/325196/16278312/bad42060-38b6-11e6-9244-c02ef25a603c.png)
This button kicks off all code generators and adds the generated sources directories as source directories, having them available during build.

But if you prefer a hard-coded string, I'll change it to that then.

Vampire commented:
Now you have it with the static string

rliesenfeld commented:
I tested this today, by running a test suite with "-javaagent:jmockit-1.8.jar" while having 1.24 in the classpath already.

No problem occurs, and the reason is that when running with -javaagent, the JVM adds the given agent jar to the *end* of the classpath. So, it even loads the "PreMain-Class" (specified in the jar's MANIFEST.MF file) from the *first* jar in the classpath, which is the desired one (ie, jmockit-1.24.jar).

As for running inside an app server, I don't know yet. I imagine the problem then is that the -javaagent jar loads from system classloader, while a jmockit.jar inside the application deployment loads from the app server loader specific to that app. At the moment, I am inclined to not do any version check.

Vampire commented:
Hm, could be about the different situation when being in app server vs. stand-alone. But I think there are also other constellations where this happens as soon as you have some framework with custom classloaders like an Eclipse-based application / plugin or an IntelliJ IDEA-based application / pluguin or something like that.

I still hope you consider adding this version check, as I think it will become essential for us, that there is not any wrong version loaded in the JVM already.

I updated the PR to current trunk.

rliesenfeld commented:
I am not comfortable applying this PR, since I can't see the problem actually happening.
If a real-world situation comes up where it does happen, and it can be shown as a reproducible failure, please open an issue.

Vampire commented:
But it IS a real-world problem that is reproducible here, please tell me what I need to do to convince you.

rliesenfeld commented:
But how can *I* reproduce the problem?

Vampire commented:
*Note:* I modified the commit to add one try-catch to better handle an already loaded JMockit that has no version field. The current commit is https://github.com/Vampire/jmockit1/commit/0c872261129155fd6d24c530c32396753ac8a4fc, you are just not seeing it in this PR, as you closed it. I guess if you reopen it, it will be assigned to this PR again, otherwise I can also open a new PR.


I wrote a little test for you to demonstrate the problem and the fix.
For this you need:

* the final JMockit 1.8 jar
* the final JMockit 1.24 jar
* and a JMockit 1.25 jar that has PR #309 (or otherwise custom classloader does not work properly anyway) and PR #302 applied.

* Take the code given below
* adapt the paths to your three jmockit JARs
* compile it against JMockit 1.25
* run it, but do **not** include JMockit in the run classpath

If you run it with argument `1`, everything will be fine, as 1.24 is used both times.
If you run it with argument `2`, you will get `java.lang.NoSuchFieldError: $MMB`
If you run it with argument `3`, you will get `JMockit with version 'null' is already loaded in this JVM, JMockit with version '1.25' could not be loaded!`

Of course this test is constructed, but it shows the problem.
If you want to test it in a real-life situation you could e. g. deploy an EAR to WildFly that uses JMockit 1.8 and after that one that uses JMockit 1.24 and then one that uses JMockit 1.25 with the two PRs applied and you will see the same results I think.

Vampire commented:
Ok, with the custom class loading support removed I was not able to construct a viable test case for this, so for now this PR is not useful anymore.
